### PR TITLE
Do not delete cache image when exporting cache if it is the same as before

### PR DESCRIPTION
**Pull request by lukasberger**
With the reproducibility improvements, it often happens that the old and
new cache images are the same. The original image should not be
deleted in this case.

This is an issue with remote images, where the image is deleted by
digest and both the old and new images share the same digest.

Signed-off-by: Lukas Berger <bergerl@google.com>

### CODE REVIEW

File: cache/image_cache.go [hunk 1]
```diff
@@ -102,10 +102,25 @@ func (c *ImageCache) Commit() error {
...
+func (c *ImageCache) DeleteOrigImage() error {
+	origIdentifier, err := c.origImage.Identifier()
+	if err != nil {
+		return errors.Wrap(err, "getting identifier for original image")
+	}
+	newIdentifier, err := c.newImage.Identifier()
+	if err != nil {
+		return errors.Wrap(err, "getting identifier for new image")
+	}
+	if origIdentifier == newIdentifier {
```
Comment (ekcasey) [inline cache/image_cache.go:122]: I think it would be safer to compare the string values here in case the identifier is a pointer.

Comment (lukasberger) [reply cache/image_cache.go:122]: Is the worry here that the signature of `Identifier()` might change to return a pointer and cause this line to always evaluate to false and the original image to be deleted? The second test added in the PR should catch that. Converting to a string has drawbacks like discarding the identifier type. However, if you feel strongly this should be a string comparison, I'm happy to change it.

Comment (ekcasey) [reply cache/image_cache.go:122]: b/c `Identifier()` returns an interface, a given implementation of the `imgutil.Image` interface can already validly return a pointer. For example if I changed your test setup from
```
fakes.NewImage("fake-image", "", local.IDIdentifier{ImageID: "fakeOrigImage"})
```
to
```
fakes.NewImage("fake-image", "", &local.IDIdentifier{ImageID: "fakeOrigImage"})
```
everything compiles but the tests fail.

I am not worried about losing the `Identifier` type information in the comparison here. We already implicitly assume that the string representation carries all necessary information when add the string value to a label or the `analyzed.toml` file, for example, and assume that string still uniquely identifies an image.

Comment (lukasberger) [reply cache/image_cache.go]: That's a very good point, changed to string comparison. I also had to update some tests because they returned nil for the identifiers which cause a panic. We could be more defensive and check for nil the `DeleteOrigImage` function, but if feels like a return value `(nil, nil)` from `Identifier()` should be invalid and was an artifact of the test setup. Thoughts?

Comment (ekcasey) [reply cache/image_cache.go]: > feels like a return value (nil, nil) from Identifier() should be invalid and was an artifact of the test setup

I agree :)
